### DISCUSSION TITLE
Waiting for revalidatePath to finish should be impossible but isn't

### DISCUSSION THREAD

**Discussion by dinosmajovic** [vercel/next.js#66267] [App Router]
Hey good people! I've been playing around with next caching trying to figure out how to await `revalidatePath` (or `revalidateTag`) and I've managed to do it even though [this](https://github.com/vercel/next.js/discussions/53206#discussioncomment-7492634) says "There isn't a way to wait for the event to finish.". This is in a component that has "use client" at the top.

```
<form
  action={(formData: FormData) => {
    // createBooking is calling revalidatePath()
    startTransition(() => createBooking(formData));

    // Waits for revalidation to complete before setting state
    setIsOpen(false);
  }}
>
```

So I guess my question is.. why does this await the revalidation but the below doesn't and is there a way to get the response value from this server action?

```
<form
  action={(formData: FormData) => {
    startTransition(async () => {
      const response = await createBooking(formData);

      // doesn't wait for revalidation to complete, just waits for the response
      if (response.ok) {
        setIsOpen(false);
        // Do something with the response
      } else {
        // Handle the error
      }
    });
  }}
>
```

**Comment by serdar-webdev** [comment]
Hi, yes you can get response from server actions but its not suggested to do, cause server actions used to mutate data and revalidate fetched data. use server fetching to get data, use server actions to mutate these data and revalidatePath or revalidateTag. 
Just use "return" in server action function;
ex : return { data : { fromServer : "hello from server" } }

**Reply by dinosmajovic** [reply]
Right, I could also use `useFormState` to get access to that return in a "proper" way, but that still doesn't answer the question: why is `revalidatePath` in the server action "awaited" in the first example, but not in the second?

**Comment by stephan-noel** [comment]
I tried this with a button without using the action prop on a form and it didn't work.  I'm on next 14.1.1.  Maybe it has something to do with async vs sync transitions and the action prop on form elements being associated with possible rerouting.  Just guessing.

I'm also curious if this is expected and if we can rely on it.
